### ISSUE TITLE
"Attempt to split row" error on nested comprehensions

### ISSUE THREAD

**Issue by wricciot** [links-lang/links#777]
The following may be related to #575.

Consider the code:
```
var foo = [] : [(i : Int, f : Int)];

for 
  (t1 <- for (t1 <- foo) [(i1 = t1.i)],
   t2 <- for (t1 <- foo) [(i1 = t1.i)])
[(i1 = t1.i1)];
```
Running this in Links causes the following IR typechecking error:
`***: Error: Links_core.TypeUtils.TypeDestructionError("Attempt to split row f:Int,i:Int on absent field i1")`

The error appears to happen when checking the body of the outer for: at that scope, `t1` should have type `[(i1 : Int)]`, but according to the environment it is `[(i : Int, f : Int)]`.

Unlike #575, this problem also happens with optimise=true, and doesn't seem to be linked to other configurations settings.

An interesting remark is that when the outer for has only one generator, the code typechecks correctly:
```
for (t1 <- for (t1 <- foo) [(i1 = t1.i)])
[(i1 = t1.i1)];

[] : [(i1:Int)]
```

**Comment by jamescheney** [comment]
What happens if you rename all the variables to eliminate the shadowing between the first and last binding of t1?

**Comment by dhil** [comment]
This is not an IR typing bug. This bug occurs during an `TransformSugar` pass.
Trace:
```
Raised at file "core/typeUtils.ml", line 8, characters 14-44
Called from file "core/typeUtils.ml", line 72, characters 6-83
Called from file "core/typeUtils.ml", line 94, characters 17-35
Called from file "core/transformSugar.ml", line 428, characters 36-65
Called from file "core/transformSugar.ml", line 651, characters 27-39
Called from file "core/transformSugar.ml", line 404, characters 36-44
Called from file "core/transformSugar.ml", line 410, characters 14-27
Called from file "core/transformSugar.ml", line 651, characters 27-39
Called from file "core/transformSugar.ml", line 111, characters 30-35
Called from file "core/transformSugar.ml", line 318, characters 27-56
Called from file "core/transformSugar.ml", line 651, characters 27-39
Called from file "core/transformSugar.ml", line 327, characters 29-37
Called from file "core/transformSugar.ml", line 651, characters 27-39
Called from file "core/transformSugar.ml", line 211, characters 40-48
Called from file "core/desugarCP.ml", line 147, characters 27-52
Called from file "core/frontend.ml", line 209, characters 15-46
Called from file "list.ml", line 121, characters 24-34
Called from file "core/frontend.ml", line 219, characters 4-116
Called from file "bin/repl.ml", line 289, characters 34-87
Called from file "bin/repl.ml", line 269, characters 44-54
Called from file "camlinternalLazy.ml", line 29, characters 17-27
Re-raised at file "camlinternalLazy.ml", line 36, characters 4-11
***: Error: Links_core.TypeUtils.TypeDestructionError("Attempt to split row f:Int,i:Int on absent field i1")
```

**Comment by wricciot** [comment]
Ah thanks. I guess there's no relationship with the other bug then.

**Comment by jamescheney** [comment]
Shadowing bug?

**Comment by wricciot** [comment]
It appears that the problem is not with shadowing as such, in the sense that the scopes of the two occurrences of t1 should be disjoint, thus shadowing must not occur.

Inspecting transformSugar.ml and desugarFors.ml reveals that when we process an Iteration, we update the environment with names declared by the generators, then move to processing the subexpressions; however, we forget to **restore the original environment** before returning.

Changing the code accordingly fixes the issue on my system. I will commit the fix as part of PR #540.
